#31 see time location schedule

KennanHoa opened the pull request:
### Related Issues
#31 

### Description
- Model, Repo modified
- ModelView, xml added/modified
- Can see list (ordered by time) of schedule that you are part of from specific group.

### How to check PR
- To check the list of schedules from a group: In database, User need to be a member of multiple schedules set by the same group.

KhangZhengNg committed:
#37 users can now accept or reject a schedule

KhangZhengNg committed:
#37 extract string resource

KhangZhengNg committed:
#37 prevent user spam respond a schedule, update Notification model

KennanHoa committed:
Merge pull request #169 from Landry333/#37-response-schedule

#37 response schedule

KennanHoa committed:
Added links to see schedule

- Button from Monitoring Group
- New schedule view
- list of schedule skeleton

KhangZhengNg committed:
#37 refactor viewModel name to be more precise

KhangZhengNg committed:
#37 simple update on UI

KennanHoa committed:
Update ListOfScheduleActivity.java

KennanHoa committed:
Merge branch '#31-see-time-location-schedule' into #37-response-schedule

KennanHoa committed:
Merge pull request #171 from Landry333/#37-response-schedule

#37 response schedule

KennanHoa committed:
#31 Schedule list working implementation

#31
- Model, Repo modified
- ModelView, xml added/mod
- Can see list (ordered by time) of schedule that you are part of from specific group.

KennanHoa committed:
#31 modified Schedule model

#31
- Added setter/getter in Schedule model

ShawnGuerra-Bautista commented:
![image](https://user-images.githubusercontent.com/23643192/101838321-68470d80-3b0e-11eb-91d9-66369d8886cc.png)

For the Group Name and Group Supervisor, the actual name and supervisor should be shown.

Other comments will be followed.

ShawnGuerra-Bautista commented:
![image](https://user-images.githubusercontent.com/23643192/101838884-63cf2480-3b0f-11eb-8cd3-068bd5ce5ff6.png)

The "Group Name" placeholder should be replaced by the actual group name, or be removed entirely.

ShawnGuerra-Bautista commented on app/src/main/res/layout/activity_supervised_group_page.xml:
```diff
@@ -1,9 +1,43 @@
+        tools:context=".activity.SupervisedGroupPageActivity">
+
+        <ImageView
+            android:id="@+id/image_view"
```
I believe this id should have a more specific name like "supervised_group_page_image_view" or something shorter but unique.

ShawnGuerra-Bautista commented on app/src/main/java/com/example/bigowlapp/repository/Repository.java:
```diff
@@ -19,8 +19,8 @@
 
-    private final FirebaseFirestore mFirebaseFirestore;
-    private final CollectionReference collectionReference;
+    protected final FirebaseFirestore mFirebaseFirestore;
+    protected final CollectionReference collectionReference;
```
Is there a reason these should be protected instead of private?

I want to hear the reasoning behind this as I don't think other classes need to access this variable directly and I don't see any class that does.

refatK replied:
I think its because they are refrenced in a child class. `ScheduleRepository`. If they are private, they can't be referenced.

ShawnGuerra-Bautista commented on app/src/main/java/com/example/bigowlapp/model/Schedule.java:
```diff
@@ -1,5 +1,149 @@
+    public static class UserResponse {
+        private Response response;
+        private Timestamp responseTime;
+        private boolean exists;
```
I believe this variable "exists" should be taken out since it isn't necessary anymore if I'm not wrong.

@refatK Please confirm this.

refatK replied:
yeah, we dpnt need it anymore

ShawnGuerra-Bautista commented on app/src/main/java/com/example/bigowlapp/activity/SupervisedGroupPageActivity.java:
```diff
@@ -1,21 +1,33 @@
+import com.google.firebase.auth.FirebaseAuth;
 
 public class SupervisedGroupPageActivity extends AppCompatActivity {
+    String groupID;
```
Should have an access modifier of private, I believe.

ShawnGuerra-Bautista commented on app/src/main/java/com/example/bigowlapp/activity/SupervisedGroupListActivity.java:
```diff
@@ -50,7 +51,12 @@ private void subscribeToData() {
+                                    supervisedGroupsListView.setOnItemClickListener((arg0, v, position, arg3) ->{
+                                        Intent intent = new Intent(getBaseContext(), SupervisedGroupPageActivity.class);
+                                        intent.putExtra("groupID", supervisedGroups.get(position).getuId());
+                                        Log.d("test", supervisedGroups.get(position).getuId());
```
This log should be taken out, I believe.

ShawnGuerra-Bautista commented on app/src/main/java/com/example/bigowlapp/activity/ListOfScheduleActivity.java:
```diff
@@ -0,0 +1,110 @@
+public class ListOfScheduleActivity extends AppCompatActivity {
+    private ListView scheduleListView;
+    private ScheduleListViewModel scheduleListViewModel;
+    String groupID;
```
Should have an access modifier of private.

ShawnGuerra-Bautista commented on app/src/main/java/com/example/bigowlapp/activity/ScheduleTestActivity.java:
```diff
@@ -0,0 +1,75 @@
+
+// TODO remove the whole activity
+public class ScheduleTestActivity extends AppCompatActivity {
+    final String channelId = "ScheduleNotificationChannelId";
+    NotificationCompat.Builder builder;
+    int nid;
```
All three variables should have an access modifier of private.

More importantly, these variables are not descriptive enough; please change their names for more descriptive ones.

ShawnGuerra-Bautista reviewed (changes requested):
Overall, I find no problem running the feature and it seems to be good. Great job.

However, I request certain changes in the code. Also, there are some UI related issues that I mentioned previously that may need changes.

Also, if you weren't aware, you should be merging with staging to avoid conflicts and to be able to push this branch to staging.

SeifElDerby commented:
the group supervisor and group name have to be modified to show the actual name also in the UI the group name should be written on top. Otherwise the feature runs perfectly.
<img width="214" alt="Screen Shot 2020-12-12 at 4 34 06 AM" src="https://user-images.githubusercontent.com/47647131/101970489-6aef5300-3c33-11eb-93e1-0d8e4a5eca9b.png">
<img width="280" alt="Screen Shot 2020-12-12 at 4 33 59 AM" src="https://user-images.githubusercontent.com/47647131/101970490-6b87e980-3c33-11eb-86b1-8cfb427601b0.png">

ShawnGuerra-Bautista commented on app/src/main/java/com/example/bigowlapp/model/Schedule.java:
```diff
@@ -1,5 +1,149 @@
+    private Timestamp endTime;
+    private GeoPoint location;
+    private String groupSupervisorUId;
+    private Map<String, UserResponse> members;
```
Naming is a little unclear and need refactoring.

The variable should be rename to userReponseMap.

refatK commented on app/src/main/java/com/example/bigowlapp/activity/AddUsersActivity.java:
```diff
@@ -18,7 +18,7 @@
 
 
-public class AddUsers extends AppCompatActivity {
+public class AddUsersActivity extends AppCompatActivity {
```
Good rename

refatK commented on app/src/main/java/com/example/bigowlapp/activity/ListOfScheduleActivity.java:
```diff
@@ -0,0 +1,110 @@
+        }
+    }
+
+    private class ScheduleAdaptor extends ArrayAdapter<Schedule> {
```
`ScheduleAdaptor` should be `ScheduleAdapter`

refatK commented on app/src/main/res/values/strings.xml:
```diff
@@ -29,4 +29,15 @@
+    <string name="end_time">End Time</string>
+    <string name="you_last">You last</string>
+    <string name="this_schedule_on">this schedule on</string>
+    <string name="textview">TextView</string>
```
This string.xml can be removed as suggest by ReviewDog

refatK commented on app/src/main/res/layout/activity_schedule_test.xml:
```diff
@@ -0,0 +1,16 @@
+        android:layout_margin="150dp"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:text="notify" />
```
should use @string resource here, as stated by reviewdog

refatK commented on app/src/main/res/layout/activity_home.xml:
```diff
@@ -99,6 +99,13 @@
+                    android:id="@+id/btn_notify_test"
+                    android:layout_width="wrap_content"
+                    android:layout_height="wrap_content"
+                    android:text="notify test" />
```
Should use android @string resource

refatK commented on app/src/main/res/layout/activity_home.xml:
```diff
@@ -99,6 +99,13 @@
                     android:text="Log Out"
                     tools:ignore="HardcodedText" />
+
+                <!-- TODO remove below button element -->
```
Will this be removed in this pr?

refatK commented on app/src/main/res/layout/activity_add_users.xml:
```diff
@@ -22,7 +22,7 @@
         android:layout_height="wrap_content"
         android:layout_marginTop="192dp"
-        android:text="This is the home page"
+        android:text="Choose invite type"
```
Should use Android @string resource

refatK commented on app/src/main/java/com/example/bigowlapp/activity/HomePageActivity.java:
```diff
@@ -23,6 +23,8 @@
 public class HomePageActivity extends AppCompatActivity {
     Button btnLogOut, btnAddUsers, btnMonitoringGroup, btnSupervisedGroup;
+    // TODO remove below button
+    Button btnNotifyTest;
```
Is this just test code, if so I guess you should remove it

refatK commented on app/src/main/java/com/example/bigowlapp/activity/ScheduleTestActivity.java:
```diff
@@ -0,0 +1,75 @@
+
+        NotificationManagerCompat notificationManager = NotificationManagerCompat.from(this);
+
+        Random notification_id = new Random();
```
fix naming conventions. Should use camelCase

refatK commented on app/src/main/java/com/example/bigowlapp/model/Notification.java:
```diff
@@ -9,8 +9,13 @@
-    private Timestamp time;
+    private Timestamp timeRead;
+    private Timestamp timeSend;
+    private Schedule.UserResponse senderResponse;
```
You should extract UserResponse to its own class instead of using it as a inner class of Schedule if its needed in this class

refatK commented on app/src/main/java/com/example/bigowlapp/viewModel/ScheduleListViewModel.java:
```diff
@@ -0,0 +1,44 @@
+
+    private final AuthRepository authRepository;
+    private final GroupRepository groupRepository;
+    private final UserRepository userRepository;
+    private final ScheduleRepository scheduleRepository;
```
These look unused. I assume they can be removed.

refatK commented on app/src/main/java/com/example/bigowlapp/viewModel/ScheduleViewRespondViewModel.java:
```diff
@@ -0,0 +1,77 @@
+
+    public boolean isOneMinuteAfterLastResponse() {
+        long userLastResponseTime = getUserResponseInSchedule().getResponseTime().toDate().getTime();
+        return now().toDate().getTime() >= (userLastResponseTime + 60000);
```
You should extract the 60 seconds as a constant (https://refactoring.com/catalog/replaceMagicLiteral.html)

refatK reviewed (changes requested):
I have changes I'm requesting mostly related to quality. You should also check the SonarCloud report for further details of potential code smells that have been added to the code.

Haven't tested running the functionality yet, but will do later.

KhangZhengNg committed:
#31 remove unused temp code

KhangZhengNg committed:
#31 refactor magic literal, rename variable

KhangZhengNg committed:
#31 extract & refactor UserScheduleResponse model, fix for ReviewDog

KennanHoa committed:
Merge branch 'staging' into #31-see-time-location-schedule

KennanHoa committed:
Fixed conflicts

KennanHoa committed:
Update top bar

KennanHoa committed:
Update ListOfScheduleActivity.java

KennanHoa committed:
#31 Changes and fixes

#31
- Shows group name in list page
- Shows group name in ViewRespondActivity
- Deleted unused codes
- Small name changes

* Missing the Supervisor name in ViewRespondActivity

refatK commented on app/src/main/java/com/example/bigowlapp/model/Notification.java:
```diff
@@ -15,7 +15,7 @@
     private Timestamp timeRead;
     private Timestamp timeSend;
-    private Schedule.UserResponse senderResponse;
+    private UserScheduleResponse senderResponse;
```
@KhangZhengNg I think `UserScheduleResponse` should just be called `UserResponse`. It's used by a Schedule, but its not necessarily related to schedule, such as here, where it is used by a generic notification.

refatK replied:
@KhangZhengNg Also I think the `exists` variable can be removed from the `UserScheduleResponse` class, assuming its unused.

refatK commented:
Tested it and everything seem to work. The group name is also correctly set now, but the supervisor name is still missing:

![image](https://user-images.githubusercontent.com/31963426/103253546-a2f2d780-494f-11eb-8d94-e86c52c27094.png)

ShawnGuerra-Bautista commented on app/src/main/java/com/example/bigowlapp/model/UserScheduleResponse.java:
```diff
@@ -0,0 +1,43 @@
+public class UserScheduleResponse {
+    private Response response;
+    private Timestamp responseTime;
+    private boolean exists;
```
This should be removed as it's a useless variable.

ShawnGuerra-Bautista replied:
The exists variable should be removed since it's a useless variable. The issue with firebase has been fixed.

ShawnGuerra-Bautista commented on app/src/main/res/layout/activity_add_users.xml:
```diff
@@ -4,13 +4,36 @@
         layout="@layout/app_bar"
         android:layout_width="match_parent"
         android:layout_height="wrap_content" />
+<!--    <ImageView-->
+<!--        android:id="@+id/imageView"-->
+<!--        android:layout_width="413dp"-->
+<!--        android:layout_height="226dp"-->
+<!--        app:layout_constraintBottom_toBottomOf="parent"-->
+<!--        app:layout_constraintEnd_toEndOf="parent"-->
+<!--        app:layout_constraintStart_toStartOf="parent"-->
+<!--        app:layout_constraintTop_toTopOf="parent"-->
+<!--        app:layout_constraintVertical_bias="0.0"-->
+<!--        app:srcCompat="@drawable/logo" />-->
+
+<!--    <TextView-->
+<!--        android:id="@+id/textView2"-->
+<!--        android:layout_width="wrap_content"-->
+<!--        android:layout_height="wrap_content"-->
+<!--        android:layout_marginTop="192dp"-->
+<!--        android:text="Choose invite type"-->
+<!--        app:layout_constraintBottom_toBottomOf="parent"-->
+<!--        app:layout_constraintEnd_toEndOf="parent"-->
+<!--        app:layout_constraintHorizontal_bias="0.501"-->
+<!--        app:layout_constraintStart_toStartOf="parent"-->
+<!--        app:layout_constraintTop_toTopOf="parent"-->
+<!--        app:layout_constraintVertical_bias="0.113" />-->
```
If these lines aren't being used, please remove them.

ShawnGuerra-Bautista commented on app/src/main/java/com/example/bigowlapp/model/Notification.java:
```diff
@@ -9,8 +9,13 @@
-    private Timestamp time;
+    private Timestamp timeRead;
+    private Timestamp timeSend;
+    private UserScheduleResponse senderResponse;
```
This is a Notification class, thus UserScheduleResponse shouldn't been in the "Notification" class as well as the other variables.

A Notification's purpose should only be used for notifications. I understand that this solution is used because you may have a problem with polymorphism and subclasses, but a preferable solution is to create a new class called UserScheduleNotification as a temporary solution.

This solution is much better since you aren't coupling and mixing up classes together. This will be a problem if Notifications such as friend requests or other types of requests are involved.

ShawnGuerra-Bautista reviewed (changes requested):
Appreciated the changes, but I noticed that the Notification class isn't being used correctly. This may cause a lot of trouble for future development.

I suggest that the Notification class be renamed to something else like UserScheduleNotification, and keep the Notification class like it was before.

This can discussed in a meeting if it's necessary.

KennanHoa committed:
Merge branch 'staging' into #31-see-time-location-schedule

KennanHoa committed:
Fixed confict

KennanHoa committed:
fixed name

refatK committed:
[#31] Added NotificationRepository polymorphism fix

Co-authored-by: Shawn Guerra-Bautista <ShawnGuerraBautista@gmail.com>

refatK committed:
[#31] Schedule Notifications are now a subclass of the base Notification class

Co-authored-by: Shawn Guerra-Bautista <ShawnGuerraBautista@gmail.com>

KennanHoa committed:
#31 Updated notification Supervisor name

#31
Fixed the name of the Supervisor in the notification page.

refatK commented:
![image](https://user-images.githubusercontent.com/31963426/104112665-abd49900-52bf-11eb-91d2-744bfa81994a.png)

Seems like there's no ids in the page anymore, so approved. Although there is a bug with the supervisor name which will be fixed as a part of #168
